External/FreeSASA/RDFreeSASA.cpp

```diff
@@ -0,0 +1,209 @@
+      return false;
+  }
+
+  bool success = true;
+  for (ROMol::AtomIterator at = mol.beginAtoms(); at != mol.endAtoms(); ++at) {
+    Atom *atom = *at;
+    int cls = -1;
+    std::string classification = "Unclassified";
+    double radius = 0.0;
+
+    const AtomMonomerInfo *info = atom->getMonomerInfo();
+    if (info) {
+      const char *atom_name = info->getName().c_str();
+      const char *res_name = 0;
+
+      if (info->getMonomerType() == AtomMonomerInfo::PDBRESIDUE) {
+        res_name = ((AtomPDBResidueInfo *)info)->getResidueName().c_str();
+        radius = classifier->radius(res_name, atom_name, classifier);
+
+        if (radius == 0.0) {
+          BOOST_LOG(rdWarningLog) << "Atom " << atom->getIdx()
+                                  << " has zero radius" << std::endl;
+        }
+
+        cls = classifier->sasa_class(res_name, atom_name, classifier);
+        if (cls == FREESASA_FAIL) {
+          std::string err = boost::str(
+              boost::format("Atom %d% failed classification") % atom->getIdx());
+          throw ValueErrorException(err);
+        } else if (cls == FREESASA_WARN) {
+          BOOST_LOG(rdWarningLog) << "Atom " << atom->getIdx()
+                                  << " could not be classified" << std::endl;
+          success = false;
+        } else {
+          classification = classifier->class2str(cls, classifier);
+        }
+      }
+    }
+
+    radii.push_back(radius);
+    atom->setProp(common_properties::Atom::SASAClass, cls);
+    atom->setProp(common_properties::Atom::SASAClassName, classification);
+  }
+
+  return success;
+}
+
+double calcSASA(const ROMol &mol, const std::vector<double> &radii,
+                const SASAOpts &opts) {
+  return calcSASA(mol, radii, -1, opts);
+}
+
+double calcSASA(const ROMol &mol, const std::vector<double> &radii, int confIdx,
+                const SASAOpts &opts) {
+  PRECONDITION(mol.getNumConformers(), "No conformers in molecule");
+  PRECONDITION(confIdx < rdcast<int>(mol.getNumConformers()),
+               "Conformer index out of range");
+  PRECONDITION(mol.getNumAtoms(), "Empty molecule");
+
+  freesasa_parameters params = freesasa_default_parameters;
+  params.n_threads = 1;
+  switch (opts.d_alg) {
+    case SASAOpts::LeeRichards:
+      params.alg = FREESASA_LEE_RICHARDS;
+      break;
+    case SASAOpts::ShrakeRupley:
+      params.alg = FREESASA_SHRAKE_RUPLEY;
+      break;
+    default:
+      throw ValueErrorException("Unknown freesasa algorithm");
+  }
+
+  // sneaky, but legal :)
+  std::vector<double> coords(mol.getNumAtoms() * 3);
+  const RDGeom::POINT3D_VECT &vec = mol.getConformer(confIdx).getPositions();
+  for (size_t i = 0; i < mol.getNumAtoms(); ++i) {
+    coords[i * 3] = vec[i].x;
+    coords[i * 3 + 1] = vec[i].y;
+    coords[i * 3 + 2] = vec[i].z;
+  }
+
+  freesasa_result *res =
+      freesasa_calc_coord(&coords[0], &radii[0], mol.getNumAtoms(), &params);
+  if (!res) return 0.0;
```

Comment:
Should probably have a better way of indicating failure than returning 0.0 as the SASA. Isn't this an exception?